libtenzir/builtins/operators/load_tcp.cpp

```diff
@@ -0,0 +1,864 @@
+          if (ec) {
+            diagnostic::warning("{}", ec.message())
+              .note("failed to accept connection")
+              .primary(args.endpoint.source)
+              .emit(diagnostics);
+            return;
+          }
+          const auto ok = handle_connection(std::move(socket));
+          if (not ok) {
+            diagnostic::warning(ok.error()).emit(diagnostics);
+          }
+        };
+        caf::anon_send(handle, caf::make_action(std::move(action)));
+      });
+  }
+
+  auto
+  read_from_connection(boost::asio::ip::tcp::socket::native_handle_type handle)
+    -> caf::result<chunk_ptr> {
+    auto connection = connections.find(handle);
+    TENZIR_ASSERT(connection != connections.end());
+    auto chunk = chunk_ptr{};
+    auto should_read = false;
+    {
+      auto lock = std::unique_lock{connection->second.mutex};
+      if (connection->second.chunks.empty()) {
+        TENZIR_ASSERT(not connection->second.rp.pending());
+        connection->second.rp
+          = self->template make_response_promise<chunk_ptr>();
+        return connection->second.rp;
+      }
+      should_read = connection->second.chunks.size()
+                    == connection_state::max_queued_chunks;
+      chunk = std::move(connection->second.chunks.front());
+      connection->second.chunks.pop();
+    }
+    if (should_read) {
+      connection->second.async_read(self);
+    }
+    return chunk;
+  }
+
+  auto write_elements(Elements elements) -> caf::result<void> {
+    if (read_rp.pending()) {
+      TENZIR_ASSERT(buffer.empty());
+      read_rp.deliver(std::move(elements));
+    }
+    buffer.push(std::move(elements));
+    if (buffer.size() < max_buffered_batches) {
+      return {};
+    }
+    return write_rps.emplace(self->template make_response_promise<void>());
+  }
+
+  auto read_elements() -> caf::result<Elements> {
+    TENZIR_ASSERT(not read_rp.pending());
+    if (not buffer.empty()) {
+      auto elements = std::move(buffer.front());
+      buffer.pop();
+      while (buffer.size() < max_buffered_batches and not write_rps.empty()) {
+        auto write_rp = std::move(write_rps.front());
+        write_rps.pop();
+        TENZIR_ASSERT(write_rp.pending());
+        write_rp.deliver();
+      }
+      return std::move(elements);
+    }
+    read_rp = self->template make_response_promise<Elements>();
+    return read_rp;
+  }
+
+  auto handle_down_msg(const caf::down_msg& msg) -> void {
+    const auto connection
+      = std::ranges::find_if(connections, [&](const auto& connection) {
+          return connection.second.pipeline_executor.address() == msg.source;
+        });
+    TENZIR_ASSERT(connection != connections.end());
+    // FIXME: It is only safe to erase this connection _after_ there is no more
+    // in-flight async read for it. Otherwise, acquiring the mutex in it can
+    // throw an exception. The connections map likely has to point to shared
+    // pointers because of this.
```

Comment:
This is a must-fix as well; there's a theoretical crash when cutting a connection immediately after starting it otherwise.